### PR TITLE
Readme Status Checker Scripts

### PR DESCRIPTION
**Pull request by jenmerritt**
### Description 

This PR is for the Readme status checker scripts that check the README.md and README.metada.json files are as per the requirements for Java Samples and are formatted correctly. The scripts will run as a GitHub Action when a Pull Request is made to either the `master` or `dev` branches, as set in the main.yml file on this branch.

The two scripts README_style_checker.py and metadata_style_checker.py can also be run locally on a sample directory.

It has been tested on my fork for PRs into the jen/test-status-checks branch (as set in the main.yml file on that branch).

### CODE REVIEW

File: .github/workflows/main.yml
Comment (JonLavi) [inline .github/workflows/main.yml]: This should be updated to `main` and we should add `dev`?

File: .github/workflows/main.yml
Comment (JonLavi) [inline .github/workflows/main.yml]: Can we change the name of the job to something more meaningful, e.g. `verify-readme`?

File: .github/workflows/main.yml
Comment (JonLavi) [inline .github/workflows/main.yml:49]: Printing out `files` and not `files_modified` here seems a bit counterintuitive, but the step `id: file_changes` might just pass the result into `files`? or does `files` include all the three commented-out categories? (modified/added/removed)?

File: .github/workflows/main.yml
Comment (JonLavi) [inline .github/workflows/main.yml]: Just missing a newline here

File: .github/workflows/main.yml
Comment (JonLavi) [inline .github/workflows/main.yml]: I think folder names in unix are traditionally written lowercase, so maybe we should change to `./scripts`

File: Scripts/CI/README_Metadata_StyleCheck/Dockerfile
Comment (JonLavi) [inline Scripts/CI/README_Metadata_StyleCheck/Dockerfile]: newline missing here as well

File: Scripts/CI/README_Metadata_StyleCheck/README_style_checker.py
Comment (JonLavi) [inline Scripts/CI/README_Metadata_StyleCheck/README_style_checker.py]: Is this list for sample names only, specific sections of the readme, or any word inside a readme?
In case of the latter, we should add KML as well, but there's probably many more that we'd add along the way

File: Scripts/CI/README_Metadata_StyleCheck/README_style_checker.py
Comment (JonLavi) [inline Scripts/CI/README_Metadata_StyleCheck/README_style_checker.py]: I can't tell if the actual number of exceptions is being used anywhere, so maybe it's possible to switch to `true`/`false`? 
I do recall there was an exit code 0/1 though, so maybe there's a technical reason for this as well.

File: Scripts/CI/README_Metadata_StyleCheck/README_style_checker.py
Comment (JonLavi) [inline Scripts/CI/README_Metadata_StyleCheck/README_style_checker.py]: Missing newline here as well

File: Scripts/CI/README_Metadata_StyleCheck/action.yml
Comment (JonLavi) [inline Scripts/CI/README_Metadata_StyleCheck/action.yml]: Should we update the email here at all?

File: Scripts/CI/README_Metadata_StyleCheck/action.yml
Comment (JonLavi) [inline Scripts/CI/README_Metadata_StyleCheck/action.yml]: Another missing newline, I won't comment on any other cases

File: Scripts/CI/README_Metadata_StyleCheck/entry.py
Comment (JonLavi) [inline Scripts/CI/README_Metadata_StyleCheck/entry.py]: I remember we discussed a solution for this. At the moment this list is repeated in `README_style_checker.py`.
Perhaps we can store it in a text file `categories.txt` and read it in in both scripts, that way there's only one source of truth?

File: Scripts/CI/README_Metadata_StyleCheck/entry.py
Comment (JonLavi) [inline Scripts/CI/README_Metadata_StyleCheck/entry.py]: Some inconsistent whitespace here, might be nice to tidy it up

File: Scripts/CI/README_Metadata_StyleCheck/entry.py
Comment (JonLavi) [inline Scripts/CI/README_Metadata_StyleCheck/entry.py]: I think it would be helpful to add some comments here as to why we add together `code1 + code2` and what the return value is used for?

File: Scripts/CI/README_Metadata_StyleCheck/metadata_style_checker.py
Comment (JonLavi) [inline Scripts/CI/README_Metadata_StyleCheck/metadata_style_checker.py]: Same here, if we could read this in from a file, it would save having to update the list in three places

File: Scripts/CI/README_Metadata_StyleCheck/metadata_style_checker.py
Comment (JonLavi) [inline Scripts/CI/README_Metadata_StyleCheck/metadata_style_checker.py]: Should we remove the commented out regex? It looks identical to the one below anyway?

File: Scripts/CI/README_Metadata_StyleCheck/metadata_style_checker.py
Comment (JonLavi) [inline Scripts/CI/README_Metadata_StyleCheck/metadata_style_checker.py]: I think if we use the output of this script https://devtopia.esri.com/runtime/common-samples/tree/master/tools/java
as the source of truth, and compare it against the README.metadata.json that is uploaded in the PR, then maybe it'd be easier to run the script and just overwrite whichever comes from the PR? It does feel a bit invasive, but there wouldn't be a way to pass this check if the metadata file isn't identical to the script output anyway? 
At this point we maybe need to consider what/where we need to make changes in case this script is stopping us from merging a metadata file as is?

**Review by JonLavi — COMMENTED** [review]
Great work on this! Definitely a very useful thing to have, and as we fine-tune it it'll save us a lot of manual labour. 
I haven't gotten around to test it myself just yet, but adding a few comments for consideration.
If we could have a central file for the list of categories, and read that in at the right place, it will make dealing with those lists a bit easier.
Also noted some other points where it might be possible to do things differently (but not necessarily better - I'll leave that up to you to decide what is worht pursuing!)

File: Scripts/CI/README_Metadata_StyleCheck/README_style_checker.py
Comment (JonLavi) [reply Scripts/CI/README_Metadata_StyleCheck/README_style_checker.py]: In hindsight, might be good to have this in a separate file as well, so that we have a central place to refer to the keywords

File: .github/workflows/main.yml
Comment (jenmerritt) [reply .github/workflows/main.yml]: Yes - helps if I push the change! Thanks

File: .github/workflows/main.yml
Comment (jenmerritt) [reply .github/workflows/main.yml:49]: yes - `files` is all three: modified / added / removed. I'll improve the commenting here.

File: Scripts/CI/README_Metadata_StyleCheck/README_style_checker.py
Comment (jenmerritt) [reply Scripts/CI/README_Metadata_StyleCheck/README_style_checker.py]: This is specifically for the "tags" section and "title". It is for when it checks casing by comparing uppercase and lowercase so KML is ok as it will equal KML (but not kml) whereas e.g. GeoPackage will not equal GEOPACKAGE or geopackage. So in this list I've picked out all the titles that don't fit this rule (e.g. Play KML Tour) and any tags identified as not fitting this rule.

Other sections are left alone, e.g. Relevant API where they contain a lot / almost exclusively these proper nouns.

I've added to the commenting here too.

Comment (JonLavi) [reply Scripts/CI/README_Metadata_StyleCheck/README_style_checker.py]: Ah now I follow! Thanks 😄

File: Scripts/CI/README_Metadata_StyleCheck/action.yml
Comment (jenmerritt) [reply Scripts/CI/README_Metadata_StyleCheck/action.yml]: I'll run this by Ting and how we can handle across each SDK

File: Scripts/CI/README_Metadata_StyleCheck/entry.py
Comment (jenmerritt) [reply Scripts/CI/README_Metadata_StyleCheck/entry.py]: Definitely, I've added this along with a few of the other improvements that we discussed to an issue in CS repo for future work to improve. I'll do the tidy ups from the code reviews and leave anything that will need more thought/time to that issue for another iteration.

File: Scripts/CI/README_Metadata_StyleCheck/metadata_style_checker.py
Comment (jenmerritt) [reply Scripts/CI/README_Metadata_StyleCheck/metadata_style_checker.py]: Definitely sounds like a good extra step towards automating - I'll add these to the notes in the issue for the next iteration, thanks 👍

File: Scripts/CI/README_Metadata_StyleCheck/README_style_checker.py
Comment (jenmerritt) [reply Scripts/CI/README_Metadata_StyleCheck/README_style_checker.py]: Yeah I think you're right. I think this is being output with the exception to show where the exception is e.g.
`1. map/display-map - Error title - README should contain title, description and image.` 
`1.` being the count. I guess because count is only 0/1 here where it compares `exception_count` to 0 it could be feasible to do true/false. But it maybe works with the exit code style? I'm hesitant to break this right now but could make a note in the CS issue if you think it'd be a good tidy up / improvement to explore?

File: .github/workflows/main.yml
Comment (jenmerritt) [reply .github/workflows/main.yml]: Good idea. Changed to `readme-metadata-stylecheck`

File: Scripts/CI/README_Metadata_StyleCheck/entry.py
Comment (jenmerritt) [reply Scripts/CI/README_Metadata_StyleCheck/entry.py]: Added in 👍

File: Scripts/CI/README_Metadata_StyleCheck/README_style_checker.py
Comment (JonLavi) [reply Scripts/CI/README_Metadata_StyleCheck/README_style_checker.py]: No worries, it's nothing major. Might be worth having a play with it, but if the errors are printed with a 'count' then that's ok too.

**Comment by jenmerritt** [comment]
Thanks for the comments @JonLavi 
I made the changes initially on my testing branch in my fork so I could check everything still ran ok (also tested locally). Then merged into this branch - and corrected the branch names!!! So now the check is also visible in this PR, as expected.

File: scripts/ci/readme_metadata_style_check/README_style_checker.py
Comment (Rachael-E) [inline scripts/ci/readme_metadata_style_check/README_style_checker.py]: minor comment but stood out for me...could we make these examples a bit more inclusive and less masculine/weaponised? :D some esri related suggestions below! 

```suggestion
    We're a mapping company. -> YES
    a map and a globe. -> NO
```

File: scripts/ci/readme_metadata_style_check/README_style_checker.py
Comment (Rachael-E) [inline scripts/ci/readme_metadata_style_check/README_style_checker.py]: there are some repeats in here from essential_headers: does that matter? just wondering if there's a way to avoid repetition.

**Review by Rachael-E — APPROVED** [review]
great work @jenmerritt. I haven't had the bandwidth to be able to test this out, but I reckon you and Simone have put it through the wringer a good amount of times now that that's a good enough test. When it goes "live" we'll be able to make tweaks as we find them. I've just put in two comments, one is more a question and doesn't block anything and the other is a suggestion for more inclusive language.

Only thought is that the new scripts directory will sit at the highest level of this repo, right? I thought it would live in the hidden .github/workflows folder, like the main.yml file and our existing gradle.yml. Is there a way the scripts directory could be hidden in the .github directory too?

File: .github/workflows/main.yml
Comment (sclaridge) [inline .github/workflows/main.yml:9]: Will we need to eventually change this when master changes to main?

**Review by sclaridge — APPROVED** [review]

File: .github/workflows/main.yml
Comment (jenmerritt) [reply .github/workflows/main.yml:9]: Yeah correct - it would just be a case of making that manual update. All that would happen if the branch was renamed and we didn't add it, is the scripts just would not run on that branch, so that should help us remember / it wouldn't cause any direct issues.

File: scripts/ci/readme_metadata_style_check/README_style_checker.py
Comment (jenmerritt) [reply scripts/ci/readme_metadata_style_check/README_style_checker.py]: Thanks I'd skimmed the detail in these - and agree these should be updated. Adding these changes now.

File: scripts/ci/readme_metadata_style_check/README_style_checker.py
Comment (jenmerritt) [reply scripts/ci/readme_metadata_style_check/README_style_checker.py]: Will add this as a refactor opportunity to the open issue, as it stands the lists are both being used and compared so will leave for now 👍

**Comment by jenmerritt** [comment]
Just coming back to this having left it for the Citra project.

Thanks @Rachael-E for the suggestion to move the files, think that's a great shout and better usability. I've made that change.

@JonLavi I'll re-request your review incase there's anything else you wanted to run over.

Otherwise once everyone has approved I'll merge them in and @Rachael-E can see how useful they are in the coming months and identify any potential issues in practice!

**Review by JonLavi — APPROVED** [review]
LGTM!

**Pull request merged**
